core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -176,26 +248,40 @@ public boolean add(T object) {
         return false;
       }
 
-      // Update the indexes.
-      for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> fieldInfo :
-          mIndexMap.entrySet()) {
-        // For this field, retrieve the value to index
-        Object fieldValue = fieldInfo.getKey().getFieldValue(object);
-        // Get the index for this field
-        ConcurrentHashMap<Object, ConcurrentHashSet<T>> index = fieldInfo.getValue();
-        ConcurrentHashSet<T> objSet = index.get(fieldValue);
-        if (objSet == null) {
-          index.putIfAbsent(fieldValue, new ConcurrentHashSet<T>());
-          objSet = index.get(fieldValue);
+      if (mIndexMapUnique != null) {
+        for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, T>> fieldInfo : mIndexMapUnique
+            .entrySet()) {
+          // For this field, retrieve the value to index
+          Object fieldValue = fieldInfo.getKey().getFieldValue(object);
+          // Get the index for this field
+          ConcurrentHashMap<Object, T> index = fieldInfo.getValue();
+          // Update the indexes.
+          index.putIfAbsent(fieldValue, object);
         }
-        if (!objSet.addIfAbsent(object)) {
-          // this call can never return false because:
-          //   a. the second-level sets in the indices are all
-          //      {@link java.util.Set} instances of unbounded space
-          //   b. We have already successfully added object on mObjects,
-          //      meaning that it cannot be already in any of the sets.
-          //      (mObjects is exactly the set-union of all the other second-level sets)
-          throw new IllegalStateException("Indexed Set is in an illegal state");
+      }
+
+      if (mIndexMapNonUnique != null) {
+        for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> fieldInfo
+            : mIndexMapNonUnique.entrySet()) {
+          // For this field, retrieve the value to index
+          Object fieldValue = fieldInfo.getKey().getFieldValue(object);
+          // Get the index for this field
+          ConcurrentHashMap<Object, ConcurrentHashSet<T>> index = fieldInfo.getValue();
+          ConcurrentHashSet<T> objSet = index.get(fieldValue);
```

Comment:
this might be a bit expensive (creating a new hashmap every time).
so maybe a better way is to combine line 273 and 274:
```java
objSet = index.putIfAbsent(fieldValue, new ConcurrentHashSet<T>());
```